### PR TITLE
enable editing of snapshots

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Tries to resolve the backend part of the issue https://github.com/rotki/rotki/issues/2219

### CODE REVIEW

File: docs/api.rst
Comment (nebolax) [inline docs/api.rst:10673]: `either` shouldn't be here

File: rotkehlchen/api/v1/schemas.py
Comment (nebolax) [inline rotkehlchen/api/v1/schemas.py:2110]: Should `is False` be here?

File: rotkehlchen/db/snapshots.py
Comment (nebolax) [inline rotkehlchen/db/snapshots.py:228]: I would split validation into 2 parts. First - for creating data structures form json and second - for validating timestamps and other properties. I would do this since you need to validate data at various steps.

File: rotkehlchen/db/snapshots.py
Comment (nebolax) [inline rotkehlchen/db/snapshots.py:269]: Logic step that is not related to the database

File: rotkehlchen/db/snapshots.py
Comment (nebolax) [inline rotkehlchen/db/snapshots.py:333]: I would move this logic outside the database since it performs complex operations (not only read/write to(from) the db). Sounds like a good logic for the `rest.py`.

File: rotkehlchen/db/snapshots.py
Comment (nebolax) [inline rotkehlchen/db/snapshots.py:343]: Why do you return a message here, is it necessary? As I see from the codebase, we usually return empty message if everything went OK.

**Review by nebolax — CHANGES_REQUESTED** [review]
Will need second review by me since I probably have not noticed everything from the first time as there are too many changes and too many comments.

Please, push changes in a separate commit.

File: docs/api.rst
Comment (LefterisJP) [reply docs/api.rst:10577]: We already have endpoints with different actions. That's fine as long as the schema validation does not become more complicated.

But it's better to add arguments to the non-get endpoints.

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:333]: The only thing it does is read & write to the db. Unless I'm missing something here

File: rotkehlchen/api/v1/schemas.py
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/schemas.py:2110]: Yeah

File: docs/api.rst
Comment (prettyirrelevant) [reply docs/api.rst:10673]: It deletes the snapshot and writes a new one provided by the user for that particular timestamp.
The issue describes how the functionality works.

Will make it more descriptive

Comment (nebolax) [reply docs/api.rst:10673]: Yeah yeah, I understand what you meant. The comment was from the reader's perspective

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:262]: I basically followed the pattern used in the `accounting/export/csv.py`

Comment (nebolax) [reply rotkehlchen/db/snapshots.py:262]: I got you. But there is a difference between them: that one is in the `accounting/export` module but this one is in the `db` module and thus is intended to work directly with database (at least in my vision). But as I see almost all other db handlers follow this pattern.

What do you think? Actually It might be okay to have it here but I would better ask Lefteris about it then.

File: rotkehlchen/api/rest.py
Comment (prettyirrelevant) [reply rotkehlchen/api/rest.py:3955]: It's never a good thing to assume that a third-party input is clean

Comment (LefterisJP) [reply rotkehlchen/api/rest.py:3955]: I love you both and happy to see discussions like this being civil :hugs: 

Checking api data should be done yeah. We should have 0 assumptions for anything coming outside of our app. At some point there may be more api consumers than just the frontend,

Comment (LefterisJP) [reply rotkehlchen/api/rest.py:3955]: Haha @nebolax you assume I read the PR. I have not. I only commented on the comment of checking data coming from outside.

So now you force me to **actually spend more time readin the commented part of the code**!!!!111eoneoneon.

Anyway ... hmm .. so looking at it indeed the lists do come from the DB, but the timestamp itself comes from the outside, right?

So if the api consumer provides a timestamp for which there is no snapshot we need to inform them about it. We can't just continue and let it fail with an exception later on?

Comment (nebolax) [reply rotkehlchen/api/rest.py:3955]: Oh, sorre :)
> you assume I read the PR. I have not

From what I see there will be no error, just empty lists are returned in case if there is no data found in the db. And my comment was about that this `if` can be redundant. So the communication problem was only about that I didn't understand how `It's never a good thing to assume that a third-party input is clean` is related here.

Comment (nebolax) [reply rotkehlchen/api/rest.py:3955]: But is absolutely not a problem that deserves long discussion ;)

Comment (LefterisJP) [reply rotkehlchen/api/rest.py:3955]: It's better to return error if no data is found for this api endpoint. So long as it's also clear in the docs I think it's better to be explicit rather than implicit.

Comment (nebolax) [reply rotkehlchen/api/rest.py:3955]: Okay, cool :+1:

File: rotkehlchen/tests/api/test_snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/tests/api/test_snapshots.py:547]: `_populate_db_with_balances` & `_populate_db_with_location_data` is only used in test_snapshots.py. 
Don't you think it's an overkill moving it to utils/factories.py?

File: rotkehlchen/tests/api/test_snapshots.py
Comment (nebolax) [reply rotkehlchen/tests/api/test_snapshots.py:615]: Perhaps I meant partial modification of the snapshot (but I don't remember after 10 days :smile: ). As I see, we don't have an option to partially modify a snapshot, so let's just forget it :)

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:271]: I might be wrong though but the `snapshots.py` module encapsulates the entirety of what's needed to import, export, delete & edit a snapshot. That means all methods particular to the snapshot manipulation. 

Due to that, only showing the needed method to the outside world i.e `.export(), import(), delete()` 
Once again, it might just be me overthinking.

Your thoughts?

Comment (nebolax) [reply rotkehlchen/db/snapshots.py:271]: Hmhmhm. You think that `snapshots.py` show to the outside world `.export()`, `.import()` and `.delete()`, right? If yes, I agree. My point is about that you we should better pass `location_data` and `balances_data` (data that is fully prepared to be written to the db) rather that some extra sources such as `Path`

File: rotkehlchen/tests/api/test_snapshots.py
Comment (nebolax) [reply rotkehlchen/tests/api/test_snapshots.py:547]: Lefteris told me once to move functions to factories in a similar situation (particularly `make_ethereum_event` and `generate_tx_entries_response`) and this why I'm mentioning it here :)

Perhaps it is okay to leave them here but there is a chance to meet Lefteris's comment about it :D

File: docs/api.rst
Comment (prettyirrelevant) [reply docs/api.rst]: The other verb that used a query arg is GET

File: docs/api.rst
Comment (yabirgb) [inline docs/api.rst:10695]: Mention that this three fields are all required

File: rotkehlchen/api/v1/schemas.py
Comment (yabirgb) [inline rotkehlchen/api/v1/schemas.py:2110]: Question. Here should the timestamp match for all the entries in the list?

Comment (yabirgb) [inline rotkehlchen/api/v1/schemas.py:2110]: Maybe a more descriptive messagee?
```suggestion
            raise ValidationError("timestamp provided {data["timestamp"]} is not the same as the one for the entries provided.")
```

And why not avoid having the timestamp in the list of `balances_snapshot` and just take it from the timestamp field?

File: rotkehlchen/db/snapshots.py
Comment (yabirgb) [inline rotkehlchen/db/snapshots.py:117]: mmmm I though that we were using 4 here but seems that this is indeed what we are respecting

File: rotkehlchen/db/snapshots.py
Comment (yabirgb) [inline rotkehlchen/db/snapshots.py:254]:
```suggestion
            deserialize_timestamp(balances_data[0]['timestamp'])
```

Comment (yabirgb) [reply rotkehlchen/db/snapshots.py:254]: But one question. You have in the schema that this is a timestamp field, do you need to do this here? Why?

File: rotkehlchen/db/snapshots.py
Comment (yabirgb) [inline rotkehlchen/db/snapshots.py:313]: Why do this and not just pass the `entry['usd_value']`? is a string already right?

Comment (yabirgb) [reply rotkehlchen/db/snapshots.py:313]: Also can raise deserialization error but this would be catched by the schema and raise validationerror so I believe is not needed. But anyway the conversion is not needed since it happens in the schema validation or am I wrong?

File: rotkehlchen/tests/api/test_snapshots.py
Comment (yabirgb) [inline rotkehlchen/tests/api/test_snapshots.py:546]: `ts_now()` returns already a timestamp

**Review by yabirgb — COMMENTED** [review]
Some questions and comments

File: docs/api.rst
Comment (prettyirrelevant) [reply docs/api.rst:10671]: It doesn't create a new one if a snapshot for the timestamp is not found.

File: rotkehlchen/api/v1/schemas.py
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/schemas.py:2110]: Yes

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:313]: This method is also used when importing snapshots as CSV.

File: rotkehlchen/db/snapshots.py
Comment (yabirgb) [inline rotkehlchen/db/snapshots.py:243]: Use `deserialize_timestamp` and handle the `DeserializationError`

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:313]: Okay

File: rotkehlchen/api/v1/schemas.py
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/schemas.py:2110]: > And why not avoid having the timestamp in the list of balances_snapshot and just take it from the timestamp field?

It'll be needed to populate the DBAssetBalance when adding the balances & location data. 
Also, it can be made a read-only field when it is being edited on the frontend

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:266]: This is quite late but what do you mean here?

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:268]: Back to this...

`import_data()` handles the data preparation and validation while `_import_data()` does the actual importing.

If I move the preparation out of `import_data()`, then I'll still have to call a method of `DBSnapshot` to handle the preparation inside `rest.py`. This is because the validation is snapshot-specific i.e it's not a generic validation per say that can be used elsewhere.

Comment (nebolax) [reply rotkehlchen/db/snapshots.py:268]: Got you. One last question :smile: Will it work if you read and validate all data in `rest.py` and pass to the database `balances_list` and `location_data_list` with this prepared data?

File: rotkehlchen/db/snapshots.py
Comment (nebolax) [reply rotkehlchen/db/snapshots.py:266]: Huh :smile:

This is the same as our other discussion. I mean that from my point it is not good to read csv in the database since database in my opinion is intended to work only with sql (no queries, csv readings, etc.)

File: rotkehlchen/db/snapshots.py
Comment (prettyirrelevant) [reply rotkehlchen/db/snapshots.py:268]: Sure sure. It'll. I'd just do that ;)

**Review by nebolax — APPROVED** [review]
LGTM! (really wanted to press this green button :smile: ).

(ofc needs approve by @LefterisJP too)

File: rotkehlchen/api/v1/resources.py
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:2389]: I assume this is a remnant/typo? 
```suggestion
```

What are you doing here?

File: rotkehlchen/api/v1/schemas.py
Comment (LefterisJP) [inline rotkehlchen/api/v1/schemas.py:2120]: Hmmm why are you re-serializing those after validating the snapshot? This does not make sense to me.
Same as in `LocationDataSnapshotSchema`.

Is it because you only use this in the edit snapshot and the edit snapshot sends them to the same thing as CSV import? But then aren't you sanitizing the input again there?

So all these marhsmallow checks are useless?

I would say that you need to (1) have a step in the import where validated and deserialized data structures are imported in the DB and (2) from here directly inject them there.

Or am I missing something ?

File: rotkehlchen/db/snapshots.py
Comment (LefterisJP) [inline rotkehlchen/db/snapshots.py:263]: The docstring makes no sense.

Describe what is returned. `List to dictionaries` is not a thing. I see you return a tuple. Tuple's first element is a bool. What is that bool supposed to be? What does the second element which is a dict supposed to contain? This is what the docstring should describe.

File: rotkehlchen/db/utils.py
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:61]: Why `Any`? Shouldn't it be `Union[str, int]` ? `Any` as a type is the same as disabling mypy checks.

File: rotkehlchen/db/utils.py
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:91]: same here

File: rotkehlchen/api/v1/resources.py
Comment (prettyirrelevant) [reply rotkehlchen/api/v1/resources.py:2390]: Because `GET` is the only verb left. I used it exporting of snapshot.

- GET with `?action=download` is for downloading a snapshot
- GET with `?action=export&path=/home/xyz` is for exporting a snapshot
- GET without `action` query param is for returning the snapshot as json response

File: rotkehlchen/api/v1/resources.py
Comment (LefterisJP) [inline rotkehlchen/api/v1/resources.py:2385]: Shouldn't thetype of this be `Optional[Literal['x', 'y', ..]]` where 'x' and 'y' are the action values?

File: rotkehlchen/api/v1/resources.py
Comment (LefterisJP) [reply rotkehlchen/api/v1/resources.py:2390]: Hmm this is quite a weird behaviour. Needs to be documented properly AND communicated to the frontend properly to make sure it's used correctly.

File: rotkehlchen/api/v1/resources.py
Comment (LefterisJP) [reply rotkehlchen/api/v1/resources.py:2390]: If this is due to mypy it should be an assertion. Or a #type:ignore 
comment explaining why it's there.

This is the wrong approach since it makes no sense to have this if check. As It should never happen. Its mere existence indicates that it's possible to happen. And then what? What happens at the else? `None` is returned.

File: rotkehlchen/db/snapshots.py
Comment (LefterisJP) [inline rotkehlchen/db/snapshots.py:296]: Since add_asset_identifiers accepts multiple identifiers I would suggest to gather all entries that start with `NFT_DIRECTIVE` here and then pass them only once to self.db.add_asset_identifiers

File: rotkehlchen/tests/api/test_snapshots.py
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_snapshots.py:365]: I saw these also while rebasing and it looked a bit odd - naming wise. Can you just name it `db`? I know we don't like short names but in this case it's better. It was the only place in the entire rebase I had to do `dbhandler.user_write()` instead of `db.user_write()`

Change for the entire file.

File: rotkehlchen/utils/snapshots.py
Comment (LefterisJP) [inline rotkehlchen/utils/snapshots.py:21]: Can you also write in the docstring what this function can return? And what exceptions it may raise? I guess no exceptions but still checking.

File: rotkehlchen/utils/snapshots.py
Comment (LefterisJP) [inline rotkehlchen/utils/snapshots.py:65]: Why do you need to return a fixed size dictionary here?

Why not use the pattern we use elsewhere in the code with tuples? The typing will also be better and will enable mypy to check the typing.

So return Tuple[str, List[Balances] List[LocatoinData], ]

Succes or failure can be determined from existence or not of one of the above. Probably the error message.

File: rotkehlchen/utils/snapshots.py
Comment (LefterisJP) [inline rotkehlchen/utils/snapshots.py:95]: mention the price for WHICH asset.
```suggestion
            f'Could not find {main_currency.symbol} price for timestamp {timestamp}. Using USD for export. '
```

**Review by LefterisJP — APPROVED** [review]
LGTM.

The failing tests are already solved in other PRs.

**Pull request merged**
